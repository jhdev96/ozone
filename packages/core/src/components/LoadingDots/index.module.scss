@import '../../scss/mixins/loading-dots/variables';


@mixin positionDots {
  display: inline-flex;

  :first-child {
    margin-right: $dot-separation;
  }
}

.loader {
  display: flex;
  flex-direction: column;
  box-sizing: border-box;

  /* Loader elements */
  .top {
    @include positionDots;
  }

  .bottom {
    @include positionDots;

    .box { 
      margin-top: $dot-separation; 
    }
  }

  .box {
    min-width: 7px;
    min-height: 7px;
    border-radius: 50%;
    transition: 0.2s opacity;
    opacity: 1;
  }
}

// Position
@each $position, $value in $positionMap {
  &.#{$position} {
    align-items: $value;
  }
}

// Theme
@each $theme, $value in $themeMap {
  .#{$theme} {
    .box {
      background-color: $value;
    }
  }
}