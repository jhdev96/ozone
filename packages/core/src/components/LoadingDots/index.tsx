import React, { Component } from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import styles from './index.module.scss';


export interface LoadingDotsProps {
  /** 
   * Sets the size of the loading dots.
   */
  dotSize?: number;
  /** 
   * Positions the loading dots relative
   * to the container component.
  */
  position?: 'left' | 'center' | 'right';
  /** 
   * Alter the color of the dots to match
   * a given theme.
   */
  theme?: ThemeType | "blue" | "red";
  /**
   * Gives the loading dots additional css
   * styling.
   */
  className?: string;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
}

interface LoadingDotsState {visibleBox: 1 | 2 | 3 | 4};

type StyleReturnType = {
  opacity: number | undefined, 
  minHeight: number | undefined, 
  minWidth: number | undefined
};

class LoadingDots extends Component<LoadingDotsProps, LoadingDotsState> {
  public static displayName = 'LoadingDots';

  public static defaultProps = {
    dotSize: 7,
    positon: 'center',
    theme: 'light'
  };

  readonly state: LoadingDotsState = {visibleBox: 1};
  animate: NodeJS.Timer = setTimeout(() => {});

  componentDidMount() {
    clearTimeout(this.animate);
    
    this.animate = setInterval(() => {
      const {visibleBox} = this.state;
      let nextBox: 1 | 2 | 3 | 4;
      
      switch (visibleBox) {
        case 1:
          nextBox = 2;
          break;
        case 2:
          nextBox = 3;
          break;
        case 3:
          nextBox = 4;
          break;
        default:
          nextBox = 1;
      }
      this.setState({visibleBox: nextBox});
    }, 250);
  }

  componentWillUnmount() {
    clearInterval(this.animate);
  }

  private setStyle = (boxNumber: number): StyleReturnType => {
    const {visibleBox} = this.state;
    const {dotSize} = this.props;

    let dotDims = {minHeight: dotSize, minWidth: dotSize};
    if (visibleBox === boxNumber) {
      return {
        opacity: 1,
        ...dotDims
      }
    } 
    return {opacity: 0, ...dotDims}
  }

  public render(): JSX.Element {
    const {position, theme, className, dataTestId} = this.props;
    return (
      <div 
        className={classNames(
          className,
          styles.loader, 
          theme     && styles[theme],
          position  && styles[position]
        )}
        data-testid={dataTestId}
      >
        <div className={styles.top}>
          <div className={styles.box} style={this.setStyle(1)} />
          <div className={styles.box} style={this.setStyle(2)} />
        </div>
        <div className={styles.bottom}>
          <div className={styles.box} style={this.setStyle(4)} />
          <div className={styles.box} style={this.setStyle(3)} />
        </div>
      </div>
    )
  }
}

export default LoadingDots;