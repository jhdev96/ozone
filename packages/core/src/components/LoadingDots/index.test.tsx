import React from "react";
import { render } from '@testing-library/react';
import LoadingDots from ".";


describe('<LoadingDots />', () => {
  it('renders 4 loading dots', () => {
    render(
      <LoadingDots />
    );

    expect(document.querySelectorAll('.box'))
      .toHaveLength(4);
  });

  it('renders loading dots in the specified position', () => {
    const {container} = render(
      <>
        <LoadingDots position="left" />
        <LoadingDots position="center" />
        <LoadingDots position="right" />
      </>
    );

    expect(container).toMatchSnapshot();
  });

  it('sets the `dotSize` prop', () => {
    render(
      <LoadingDots dotSize={5} />
    );

    const box = document.querySelector('.box') as HTMLElement;

    expect(box.style.minHeight).toBe("5px");
    expect(box.style.minWidth).toBe("5px");
  });
});