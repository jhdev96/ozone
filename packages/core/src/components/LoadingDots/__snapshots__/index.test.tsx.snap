// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadingDots /> renders loading dots in the specified position 1`] = `
<div>
  <div
    class="loader light left"
  >
    <div
      class="top"
    >
      <div
        class="box"
        style="opacity: 1; min-height: 7px; min-width: 7px;"
      />
      <div
        class="box"
        style="opacity: 0; min-height: 7px; min-width: 7px;"
      />
    </div>
    <div
      class="bottom"
    >
      <div
        class="box"
        style="opacity: 0; min-height: 7px; min-width: 7px;"
      />
      <div
        class="box"
        style="opacity: 0; min-height: 7px; min-width: 7px;"
      />
    </div>
  </div>
  <div
    class="loader light center"
  >
    <div
      class="top"
    >
      <div
        class="box"
        style="opacity: 1; min-height: 7px; min-width: 7px;"
      />
      <div
        class="box"
        style="opacity: 0; min-height: 7px; min-width: 7px;"
      />
    </div>
    <div
      class="bottom"
    >
      <div
        class="box"
        style="opacity: 0; min-height: 7px; min-width: 7px;"
      />
      <div
        class="box"
        style="opacity: 0; min-height: 7px; min-width: 7px;"
      />
    </div>
  </div>
  <div
    class="loader light right"
  >
    <div
      class="top"
    >
      <div
        class="box"
        style="opacity: 1; min-height: 7px; min-width: 7px;"
      />
      <div
        class="box"
        style="opacity: 0; min-height: 7px; min-width: 7px;"
      />
    </div>
    <div
      class="bottom"
    >
      <div
        class="box"
        style="opacity: 0; min-height: 7px; min-width: 7px;"
      />
      <div
        class="box"
        style="opacity: 0; min-height: 7px; min-width: 7px;"
      />
    </div>
  </div>
</div>
`;
